File: Backend/Services/DashboardService.cs
using Backend.Data;
using Backend.DTOs.Dashboard;
using Backend.Models.Enums;
using Backend.Services.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace Backend.Services;

public class DashboardService : IDashboardService
{
    private readonly AppDbContext _context;

    public DashboardService(AppDbContext context)
    {
        _context = context;
    }

    public async Task<DashboardKpisDto> ObterKpisAsync()
    {
        var hoje = DateTime.UtcNow.Date;
        var amanha = hoje.AddDays(1);

        // Consultas otimizadas - executadas em paralelo
        var receitaTotalTask = _context.Pedidos
            .Where(p => p.Status == StatusPedido.Entregue)
            .SumAsync(p => (decimal?)p.ValorTotal) ?? Task.FromResult<decimal?>(0);

        var totalPedidosTask = _context.Pedidos.CountAsync();
        var totalClientesTask = _context.Clientes.CountAsync();

        var pedidosPendentesTask = _context.Pedidos
            .CountAsync(p => p.Status == StatusPedido.Pendente || 
                           p.Status == StatusPedido.EmProducao || 
                           p.Status == StatusPedido.Pronto);

        var pedidosHojeTask = _context.Pedidos
            .CountAsync(p => p.DataCriacao >= hoje && p.DataCriacao < amanha);

        var receitaHojeTask = _context.Pedidos
            .Where(p => p.DataCriacao >= hoje && 
                       p.DataCriacao < amanha && 
                       p.Status == StatusPedido.Entregue)
            .SumAsync(p => (decimal?)p.ValorTotal) ?? Task.FromResult<decimal?>(0);

        await Task.WhenAll(receitaTotalTask, totalPedidosTask, totalClientesTask, 
                          pedidosPendentesTask, pedidosHojeTask, receitaHojeTask);

        return new DashboardKpisDto
        {
            ReceitaTotal = await receitaTotalTask ?? 0,
            TotalPedidos = await totalPedidosTask,
            TotalClientes = await totalClientesTask,
            PedidosPendentes = await pedidosPendentesTask,
            PedidosHoje = await pedidosHojeTask,
            ReceitaHoje = await receitaHojeTask ?? 0
        };
    }

    public async Task<List<PedidosPorMesDto>> ObterPedidosPorMesAsync(int meses = 6)
    {
        var dataInicio = DateTime.UtcNow.AddMonths(-meses + 1).Date;
        dataInicio = new DateTime(dataInicio.Year, dataInicio.Month, 1);

        var dados = await _context.Pedidos
            .Where(p => p.DataCriacao >= dataInicio)
            .GroupBy(p => new { p.DataCriacao.Year, p.DataCriacao.Month })
            .Select(g => new
            {
                Ano = g.Key.Year,
                Mes = g.Key.Month,
                TotalPedidos = g.Count(),
                ReceitaTotal = g.Where(p => p.Status == StatusPedido.Entregue).Sum(p => p.ValorTotal)
            })
            .OrderBy(x => x.Ano)
            .ThenBy(x => x.Mes)
            .ToListAsync();

        var nomesMeses = new[] { "", "Jan", "Fev", "Mar", "Abr", "Mai", "Jun", 
                                  "Jul", "Ago", "Set", "Out", "Nov", "Dez" };

        return dados.Select(d => new PedidosPorMesDto
        {
            Ano = d.Ano,
            Mes = d.Mes,
            MesNome = nomesMeses[d.Mes],
            TotalPedidos = d.TotalPedidos,
            ReceitaTotal = d.ReceitaTotal
        }).ToList();
    }

    public async Task<List<DistribuicaoStatusDto>> ObterDistribuicaoStatusAsync()
    {
        var total = await _context.Pedidos.CountAsync();
        if (total == 0)
        {
            return new List<DistribuicaoStatusDto>();
        }

        var distribuicao = await _context.Pedidos
            .GroupBy(p => p.Status)
            .Select(g => new
            {
                Status = g.Key,
                Quantidade = g.Count()
            })
            .ToListAsync();

        return distribuicao.Select(d => new DistribuicaoStatusDto
        {
            Status = FormatarStatus(d.Status),
            Quantidade = d.Quantidade,
            Percentual = Math.Round((decimal)d.Quantidade / total * 100, 2)
        }).ToList();
    }

    public async Task<DashboardCompletoDto> ObterDashboardCompletoAsync()
    {
        var kpisTask = ObterKpisAsync();
        var pedidosPorMesTask = ObterPedidosPorMesAsync();
        var distribuicaoStatusTask = ObterDistribuicaoStatusAsync();

        await Task.WhenAll(kpisTask, pedidosPorMesTask, distribuicaoStatusTask);

        return new DashboardCompletoDto
        {
            Kpis = await kpisTask,
            PedidosPorMes = await pedidosPorMesTask,
            DistribuicaoStatus = await distribuicaoStatusTask
        };
    }

    private static string FormatarStatus(StatusPedido status) => status switch
    {
        StatusPedido.Pendente => "Pendente",
        StatusPedido.EmProducao => "Em Produção",
        StatusPedido.Pronto => "Pronto",
        StatusPedido.EmEntrega => "Em Entrega",
        StatusPedido.Entregue => "Entregue",
        StatusPedido.Cancelado => "Cancelado",
        _ => status.ToString()
    };
}

File: backend/src/services/dashboard.service.ts
import { supabase } from '../config/database';
import { StatusPedido, StatusPedidoLabel } from '../models/enums';
import {
  DashboardKpisDto,
  PedidosPorMesDto,
  DistribuicaoStatusDto,
  DashboardCompletoDto,
} from '../dtos/dashboard.dto';

// ─── KPIs ────────────────────────────────────────────────────────────
export async function obterKpisAsync(): Promise<DashboardKpisDto> {
  const hoje = new Date();
  hoje.setUTCHours(0, 0, 0, 0);
  const amanha = new Date(hoje);
  amanha.setUTCDate(amanha.getUTCDate() + 1);

  // Executar consultas em paralelo para performance
  const [pedidosResult, clientesResult, pedidosHojeResult] = await Promise.all([
    // Todos os pedidos (status + valor)
    supabase.from('pedidos').select('status, valor_total, data_criacao'),
    // Total de clientes
    supabase.from('clientes').select('id', { count: 'exact', head: true }),
    // Pedidos criados hoje
    supabase
      .from('pedidos')
      .select('status, valor_total')
      .gte('data_criacao', hoje.toISOString())
      .lt('data_criacao', amanha.toISOString()),
  ]);

  const pedidos = pedidosResult.data || [];
  const totalPedidos = pedidos.length;
  const totalClientes = clientesResult.count || 0;

  // Receita total (apenas pedidos entregues)
  const receitaTotal = pedidos
    .filter((p: any) => p.status === StatusPedido.Entregue)
    .reduce((sum: number, p: any) => sum + Number(p.valor_total), 0);

  // Pedidos pendentes (Pendente + EmProducao + Pronto)
  const pedidosPendentes = pedidos.filter((p: any) =>
    [StatusPedido.Pendente, StatusPedido.EmProducao, StatusPedido.Pronto].includes(p.status),
  ).length;

  const pedidosHoje = pedidosHojeResult.data || [];
  const receitaHoje = pedidosHoje
    .filter((p: any) => p.status === StatusPedido.Entregue)
    .reduce((sum: number, p: any) => sum + Number(p.valor_total), 0);

  return {
    receitaTotal,
    totalPedidos,
    totalClientes,
    pedidosPendentes,
    pedidosHoje: pedidosHoje.length,
    receitaHoje,
  };
}

// ─── Pedidos por Mês ─────────────────────────────────────────────────
export async function obterPedidosPorMesAsync(meses: number = 6): Promise<PedidosPorMesDto[]> {
  const dataInicio = new Date();
  dataInicio.setUTCMonth(dataInicio.getUTCMonth() - meses + 1);
  dataInicio.setUTCDate(1);
  dataInicio.setUTCHours(0, 0, 0, 0);

  const nomesMeses = ['', 'Jan', 'Fev', 'Mar', 'Abr', 'Mai', 'Jun', 'Jul', 'Ago', 'Set', 'Out', 'Nov', 'Dez'];

  const { data: pedidos, error } = await supabase
    .from('pedidos')
    .select('data_criacao, status, valor_total')
    .gte('data_criacao', dataInicio.toISOString());

  if (error) throw new Error(error.message);

  // Agrupar por ano/mês no lado do cliente
  const grupoMap = new Map<string, { ano: number; mes: number; totalPedidos: number; receitaTotal: number }>();

  for (const p of pedidos || []) {
    const data = new Date(p.data_criacao);
    const ano = data.getUTCFullYear();
    const mes = data.getUTCMonth() + 1;
    const key = `${ano}-${mes}`;

    if (!grupoMap.has(key)) {
      grupoMap.set(key, { ano, mes, totalPedidos: 0, receitaTotal: 0 });
    }
    const grupo = grupoMap.get(key)!;
    grupo.totalPedidos++;
    if (p.status === StatusPedido.Entregue) {
      grupo.receitaTotal += Number(p.valor_total);
    }
  }

  return Array.from(grupoMap.values())
    .sort((a, b) => a.ano - b.ano || a.mes - b.mes)
    .map((d) => ({
      ano: d.ano,
      mes: d.mes,
      mesNome: nomesMeses[d.mes] || '',
      totalPedidos: d.totalPedidos,
      receitaTotal: d.receitaTotal,
    }));
}

// ─── Distribuição de Status ──────────────────────────────────────────
export async function obterDistribuicaoStatusAsync(): Promise<DistribuicaoStatusDto[]> {
  const { data: pedidos, error } = await supabase
    .from('pedidos')
    .select('status');

  if (error) throw new Error(error.message);

  const total = (pedidos || []).length;
  if (total === 0) return [];

  // Agrupar por status no lado do cliente
  const statusCount = new Map<number, number>();
  for (const p of pedidos || []) {
    statusCount.set(p.status, (statusCount.get(p.status) || 0) + 1);
  }

  return Array.from(statusCount.entries()).map(([status, quantidade]) => ({
    status: StatusPedidoLabel[status as StatusPedido] || status.toString(),
    quantidade,
    percentual: Math.round((quantidade / total) * 10000) / 100,
  }));
}

// ─── Dashboard Completo ──────────────────────────────────────────────
export async function obterDashboardCompletoAsync(): Promise<DashboardCompletoDto> {
  const [kpis, pedidosPorMes, distribuicaoStatus] = await Promise.all([
    obterKpisAsync(),
    obterPedidosPorMesAsync(),
    obterDistribuicaoStatusAsync(),
  ]);

  return { kpis, pedidosPorMes, distribuicaoStatus };
}
